google-cloud-clients/google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/Publisher.java

```diff
@@ -124,8 +129,10 @@ private Publisher(Builder builder) throws IOException {
 
     // Publisher used to take maxAttempt == 0 to mean infinity, but to GAX it means don't retry.
     // We post-process this here to keep backward-compatibility.
+    // Also, if "message ordering" is enabled, the publisher should retry sending the failed
+    // message infinitely rather than sending the next one.
     RetrySettings retrySettings = builder.retrySettings;
-    if (retrySettings.getMaxAttempts() == 0) {
+    if (retrySettings.getMaxAttempts() == 0 || builder.enableMessageOrdering) {
       retrySettings = retrySettings.toBuilder().setMaxAttempts(Integer.MAX_VALUE).build();
```

Comment:
There is a total rpc timeout of 10 sec in addition to max attempts which will cause the rpc to stop retrying. RPC will also fail on non retryable error codes how that should be handled? Also doesnt isnt it to fail all pending batch for this ordering key rather than retrying indefinitely and hanging application thread if there are more concurrent rpcs in progress than number of gax thread.

Reply:
I think some kind of interface providing feedback to application about the failure and let application take the decision whether to continue or to stop publishing for that ordering key would be better, rather than keep retrying for all error code indefinitely.